controller/resource.go

```diff
@@ -0,0 +1,222 @@
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+)
+
+const meshTag = "consul.hashicorp.com/mesh"
+
+// ResourceLister is an interface for listing Resources.
+type ResourceLister interface {
+	List() ([]Resource, error)
+}
+
+// Resource is a generic type that needs to be reconciled by the Controller.
+// It offers Upsert and Delete functions to reconcile itself with an external state.
+type Resource interface {
+	ID() (string, error)
+	Upsert() error
+	Delete() error
+}
+
+// TaskLister is an implementation of ResourceLister that lists ECS tasks.
+type TaskLister struct {
+	// ECSClient is the AWS ECS client to be used by the TaskLister.
+	ECSClient ecsiface.ECSAPI
+	// SecretsManagerClient is the AWS Secrets Manager client to be used by the TaskLister.
+	SecretsManagerClient secretsmanageriface.SecretsManagerAPI
+	// ConsulClient is the Consul client to be used by the TaskLister.
+	// TaskLister doesn't need to talk to Consul, but it passes this client
+	// to each Resource it creates.
+	ConsulClient *api.Client
+
+	// Cluster is the name or the ARN of the ECS cluster.
+	Cluster string
+	// SecretPrefix is the prefix to determine names of resources in Consul or AWS.
+	SecretPrefix string
+
+	// Log is the logger for the TaskLister.
+	Log hclog.Logger
+}
+
+// List lists all tasks for the Cluster.
+func (t TaskLister) List() ([]Resource, error) {
+	var resources []Resource
+	// nextToken is to handle paginated responses from AWS.
+	var nextToken *string
+
+	// This isn't an infinite loop, instead this is a "do while" loop
+	// because we'll break out of it as soon as nextToken is nil.
+	for {
+		taskListOutput, err := t.ECSClient.ListTasks(&ecs.ListTasksInput{
+			Cluster:   aws.String(t.Cluster),
+			NextToken: nextToken,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("listing tasks: %w", err)
+		}
+		nextToken = taskListOutput.NextToken
+
+		tasks, err := t.ECSClient.DescribeTasks(&ecs.DescribeTasksInput{
+			Cluster: aws.String(t.Cluster),
+			Tasks:   taskListOutput.TaskArns,
+			Include: []*string{aws.String("TAGS")},
+		})
+		if err != nil {
+			return nil, fmt.Errorf("describing tasks: %w", err)
+		}
+		for _, task := range tasks.Tasks {
+			resources = append(resources, &Tasks{
+				SecretsManagerClient: t.SecretsManagerClient,
+				ConsulClient:         t.ConsulClient,
+				Cluster:              t.Cluster,
+				Log:                  t.Log,
+				SecretPrefix:         t.SecretPrefix,
+				Task:                 task,
+			})
+		}
+		if nextToken == nil {
+			break
+		}
+	}
+	return resources, nil
+}
+
+type Tasks struct {
+	SecretsManagerClient secretsmanageriface.SecretsManagerAPI
+	ConsulClient         *api.Client
+
+	Cluster      string
+	SecretPrefix string
+	Task         *ecs.Task
+
+	Log hclog.Logger
+}
+
+// ID returns Task family/Consul service name or error if it cannot be determined from the Task group.
+// It is derived from the group which looks like "service:<service name>".
```

Comment:
Hmm, non-blocking: The [docs](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-groups.html) mention the Task group could be entirely custom (which seems uncommon). Would parsing the family from the `TaskDefinitionArn` be more reliable, since it's a known format? (I don't see another field in the Task that includes the family ☹️ )

Reply:
That's a good point. And actually, from the docs it seems that if you have a service, then the group name must be the service name. In our case, the service name is not guaranteed to be equal to the task family name. I'll change it to parse it from the task def ARN, thanks for doing the research!!

Reply:
🤔 Thinking about this more, I'm thinking to switch to tracking state with the whole Task Definition ARN instead of task family. The reason for that is that task def is mutable, and so if you, for example, update your task from not being on the mesh to now use the mesh-task, we won't detect that since we'll think that we've already "seen" this task family and won't know that it now has the mesh tag.

Because the task definition ARN will also include the revision version, we can therefore detect that the task definition has changed from the one we "know" about and should process it again.

Reply:
True!! Basically, if any version of a TaskDefinition is tagged as a mesh task, the controller should ensure an ACL token is provisioned for that TaskDefinition family.

Is that right? We could optimize slightly based on that: once an ACL token is provisioned for the TaskDefinition family, we don't need to examine any other versions of that TaskDefinition (non-blocking, for sure)

Reply:
> once an ACL token is provisioned for the TaskDefinition family, we don't need to examine any other versions of that TaskDefinition

Yeah, good idea! I think that's worth investigating.